arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Option.kt

```diff
@@ -812,19 +828,149 @@ fun <A> Iterable<Option<A>>.combineAll(MA: Monoid<A>): Option<A> =
+ */
+fun <T> Collection<T>.singleOrNone(): Option<T> =
+  when (size) {
+    1 -> firstOrNone()
+    else -> None
+  }
+
+/**
+ * Returns the single element as [Some(element)][Some] matching the given [predicate], or [None] if element was not found or more than one element was found.
+ */
+inline fun <T> Iterable<T>.singleOrNone(predicate: (T) -> Boolean): Option<T> {
+  val list = mutableListOf<T>()
+  for (element in this) {
+    if (predicate(element)) {
+      if (list.isNotEmpty()) {
+        return None
+      }
+      list.add(element)
+    }
+  }
+  return list.firstOrNone()
+}
 
-inline fun <T> Iterable<T>.firstOrNone(predicate: (T) -> Boolean): Option<T> = this.firstOrNull(predicate).toOption()
+/**
+ * Returns the last element as [Some(element)][Some], or [None] if the iterable is empty.
+ */
+fun <T> Iterable<T>.lastOrNone(): Option<T> =
+  when (this) {
+    is Collection -> lastOrNone()
+    else -> iterator().run {
+      if (hasNext()) {
+        var last: T
+        do last = next() while (hasNext())
+        Some(last)
+      } else {
+        None
+      }
+    }
+  }
 
-fun <T> Iterable<T>.singleOrNone(): Option<T> = this.singleOrNull().toOption()
+/**
+ * Returns the last element as [Some(element)][Some], or [None] if the collection is empty.
+ */
+fun <T> Collection<T>.lastOrNone(): Option<T> =
+  if (isEmpty()) {
+    None
+  } else {
+    Some(last())
+  }
 
-inline fun <T> Iterable<T>.singleOrNone(predicate: (T) -> Boolean): Option<T> = this.singleOrNull(predicate).toOption()
+/**
+ * Returns the last element as [Some(element)][Some] matching the given [predicate], or [None] if no such element was found.
+ */
+inline fun <T> Iterable<T>.lastOrNone(predicate: (T) -> Boolean): Option<T> {
+  val list = mutableListOf<T>()
+  for (element in this) {
+    if (predicate(element)) {
+      if (list.isEmpty()) {
+        list.add(element)
+      } else {
+        list[0] = element
+      }
+    }
+  }
+  return list.firstOrNone()
+}
 
-fun <T> Iterable<T>.lastOrNone(): Option<T> = this.lastOrNull().toOption()
+/**
+ * Returns an element as [Some(element)][Some] at the given [index] or [None] if the [index] is out of bounds of this iterable.
+ */
+fun <T> Iterable<T>.elementAtOrNone(index: Int): Option<T> =
+  when {
+    index < 0 -> None
+    this is Collection -> elementAtOrNone(index)
+    else -> iterator().skip(index).nextOrNone()
+  }
 
-fun <T> Iterable<T>.lastOrNone(predicate: (T) -> Boolean): Option<T> = this.lastOrNull(predicate).toOption()
+/**
+ * Returns an element as [Some(element)][Some] at the given [index] or [None] if the [index] is out of bounds of this collection.
+ */
+fun <T> Collection<T>.elementAtOrNone(index: Int): Option<T> =
+  when (index) {
+    in 0 until size -> Some(elementAt(index))
+    else -> None
+  }
 
-fun <T> Iterable<T>.elementAtOrNone(index: Int): Option<T> = this.elementAtOrNull(index).toOption()
+private tailrec fun <T> Iterator<T>.skip(count: Int): Iterator<T> =
```

Comment:
Not a blocker, but is it worth making skip public? Could be handy for someone. :slightly\_smiling\_face:

Reply:
I think it looks more like a utility function. Also, I don't think it's a good idea to add extension functions to the `Iterator` (as a public API), at least I haven't seen cases like this. :thinking:

Reply:
Actually, if we call it `drop(n)` it would match the one for Iterables in `stdlib`:  
[https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/drop.html](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/drop.html)

I think there is a use case to do things like: `iterator().drop(3).nextOrNone()` but it's a new function so probably out of the scope of this PR :)